### PR TITLE
Really efficeint representation of PI

### PR DESCRIPTION
**Pull request by KrzysztofCwalina**
I was doing research what our developers need in terms of APIs. Today, I noticed a significant spike in the demand for a good representation of the number PI. After a review of the available representations, I concluded we can do better. This PR introduces the most efficient PI representation in the history of computers. If we don't win a Turing award for this, I don't know who will.

Needless to say, I would like to see this type in the Net Standard.

### CODE REVIEW

**Comment by johanste** [comment]
I personally find the code overly complicated. If you insist on not taking a dependency on a third party package (because having a broken pi would be bad), would something like this:

```python
from decimal import Decimal, getcontext
getcontext().prec=500
pi =sum(1/Decimal(16)**k * 
          (Decimal(4)/(8*k+1) - 
           Decimal(2)/(8*k+4) - 
           Decimal(1)/(8*k+5) -
           Decimal(1)/(8*k+6)) for k in range(500))
print(pi)
```

be good enough? Of course, precision and iteration count would need to be parameterized - this is just to illustrate the basic idea...

**Comment by KrzysztofCwalina** [comment]
@johanste, I measure perf of your implementation and get back to you. Hopefully I can get the spaces and tabs in the right places.

File: PI/Azure.Core.PI/PI.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    {
+        public static bool operator ==(PI left, PI right) => true;
+        public static bool operator !=(PI left, PI right) => false;
+
+        public override string ToString() => "3.14...";
+
+        public PIDigits Digits => new PIDigits();
+    }
+
+    public static class NumerocsExtensions
```
Comment (Tornhoof) [inline PI/Azure.Core.PI/PI.cs]: Typo Numerocs->Numerics

**Comment by JonathanGiles** [comment]
If you'd like a faster implementation I'd be happy to rewrite this for you in Java :-)

**Comment by terrajobst** [comment]
Unless it's written in rust, my confidence in the code is low. Alternatively, I'd accept a functional implementation in F#.

File: PI/Azure.Core.PI/PI.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    // diameter of a circle
+    public ref struct D
+    { }
+
+    public class PIDigits
+    {
+        int _digit;
+        CancellationToken _cancellation;
+
+        public async ValueTask<bool> MoveNextAsync() {
```
Comment (terrajobst) [inline PI/Azure.Core.PI/PI.cs]: Your bracing style is inconsistent.

Comment (KrzysztofCwalina) [reply PI/Azure.Core.PI/PI.cs]: Do you prefer fast or consistent? I had to pick my battles.

**Comment by KrzysztofCwalina** [comment]
@JonathanGiles, is Java :-) (pronounced Java-colon-hyphen-parenthesis) a new variant of Java? I have not heard of it.

**Comment by KrzysztofCwalina** [comment]
@terrajobst, it won't work in functional languages like F#. Computing very large number of PI digits requires side effects. RAM is not large enough to store all the digits. 

It won't work in Rust either. See [stack overflow](https://stackoverflow.com/questions/31208465/pi-constant-is-ambiguous/31210454).

**Comment by KrzysztofCwalina** [comment]
Closing as the demand for PI has mysteriously disappeared.

**Pull request closed without merging**